### ISSUE TITLE
comparing to 0 literal

### ISSUE THREAD

**Issue by stanmihai4** [nlohmann/json#414]
Hello,

Comparing to the 0 literal results calls `operator==(const_reference v, std::nullptr_t)` since converting to a standard type is preferred to a custom one (json).

```cpp
// don't let catch decompose the expression
CHECK((json(0) == 0));
CHECK((0 == json(0)));
```

**Comment by stanmihai4** [comment]
The proposed fix fails to compiles ```json(0) == 0L```

I guess this could be fixed with CompatibleNumberIntegerType, but the constructors are already handling all those cases, including converting a nullptr.

Just removing the `operator==(const_reference v, std::nullptr_t)` seems to be a  better solution.

**Comment by nlohmann** [comment]
When `operator==(const_reference v, std::nullptr_t)` is removed, then `j == nullptr` does not work for a JSON value `j`.

**Comment by stanmihai4** [comment]
The only errors after removing the `nullptr` equality operator are on clang `error: use of overloaded operator '==' is ambiguous`

**Comment by pboettch** [comment]
When comparing `json(false)` to `false the default GCC (5.4.0) of Ubuntu 16.04  (gcc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609) also calls `bool operator==(const_reference v, std::nullptr_t)`:

```C++
json j = false;
bool b = false;

std::cerr << "j == false : " << (j == false) << "\n";
std::cerr << "j == b : " << (j == b) << "\n";
```
Gives
```
j == false : 0
j == b : 1
```

Not so on gcc 6.2, but maybe others are affected as well.

**Comment by nlohmann** [comment]
@pboettch I can confirm your issue with macOS Sierra's Clang.

**Comment by nlohmann** [comment]
Ideally would be a function like

```cpp
template<class CompatibleType, typename std::enable_if<
             std::is_constructible<basic_json, CompatibleType>::value, int>::type = 0>
friend bool operator==(CompatibleType lhs, const_reference rhs) noexcept
{
    return (basic_json(lhs) == rhs);
}
```

which basically like "comparability" to "constructability".

**Comment by stanmihai4** [comment]
Using `std::is_constructible` leads to ambiguous overloads on clang for custom classes that implement the comparison operator for json.

I updated the pull request to use `std::is_scalar` and the test to check for comparing to nullptr, int, unsigned, long, double and string literals.

Note that some of the added tests were previously generating compile errors due to ambiguous overloads, so for example `json(1) == 1` was not compiling on clang.

**Comment by nlohmann** [comment]
Merged with 057b1e606b0f3c1da222c4e0094c20d7f4aa426e. Thanks everybody!
